app/code/Magento/Catalog/Model/Product/Copier.php

```diff
@@ -76,20 +76,14 @@ public function copy(Product $product)
         $duplicate->setStoreId(\Magento\Store\Model\Store::DEFAULT_STORE_ID);
 
         $this->copyConstructor->build($product, $duplicate);
-        $isDuplicateSaved = false;
-        do {
-            $urlKey = $duplicate->getUrlKey();
-            $urlKey = preg_match('/(.*)-(\d+)$/', $urlKey, $matches)
-                ? $matches[1] . '-' . ($matches[2] + 1)
-                : $urlKey . '-1';
-            $duplicate->setUrlKey($urlKey);
-            try {
-                $duplicate->save();
-                $isDuplicateSaved = true;
-            } catch (\Magento\Framework\Exception\AlreadyExistsException $e) {
-            } catch (\Magento\UrlRewrite\Model\Exception\UrlAlreadyExistsException $e) {
-            }
-        } while (!$isDuplicateSaved);
+
+        $urlKey = $duplicate->getUrlKey();
+        $urlKey = preg_match('/(.*)-(\d+)$/', $urlKey, $matches)
+            ? $matches[1] . '-' . ($matches[2] + 1)
+            : $urlKey . '-1';
+        $duplicate->setUrlKey($urlKey);
+        $duplicate->save();
```

Comment:
Such change is incorrect as new code tries to save duplicate only once - while the whole point of loop was in guaranteed saving of a duplicated product.

@hostep suggestion makes perfect sense to me, please try to do a backport of this logic and check if it solves your problem.

Reply:
I agree with @oleksii-lisovyi here, this do while loop can cause [major problems](https://github.com/magento/magento2/issues/9466), this should be solved in a better way then doing a potential infinite loop.

Reply:
@oleksii-lisovyi @hostep 

> } catch (\Magento\Framework\Exception\AlreadyExistsException $e) {	
> } catch (\Magento\UrlRewrite\Model\Exception\UrlAlreadyExistsException $e) {

Such exceptions are specific enough, if some third-party code throws them permanently - it really needs to be fixed.

The problem mentioned is solved in 2.3 and loop is still there: https://github.com/magento/magento2/blob/50592b4fb359b1daba9de7b554c9c8afa1d8d160/app/code/Magento/Catalog/Model/Product/Copier.php#L92

> Finding a unique url key should not be done by trying to save a product over and over again, this should be done separately somehow.

Please tell more about this "somehow". Main purpose of this code is to avoid race condition (when some URL Key we would use was added by somebody else). Otherwise we would just try the next free URL Key and it would work.

And again, 2.2 should not live on its own, any backportable fix from 2.3 must be backported.

Reply:
@hostep why? This code simply works in 2.3 and I see no issue in it for 2.2

Reply:
Because a save method can trigger things you don't want to trigger more then once.
In the case of https://github.com/magento/magento2/issues/9466, calling save more then once copies the image files on the filesystem associated with the product more then once, even though by duplicating a product, you expect the image files to only be copied a single time.

Just some wild idea: can we not first add a completely random unique bunch of characters (a hash if you will) as a url key to the duplicated product, then save the product, and then generate a more sensible url key using that do while loop and only save that on the product so only a single attribute gets changed?

Reply:
@hostep if things are added more than once when calling `save`, they are not implemented well.